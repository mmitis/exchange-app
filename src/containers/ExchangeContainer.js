//@flow
import React from "react";
import {connect} from "react-redux";
import Big from 'big.js';
import {bindActionCreators} from "redux";
import isEqual from 'lodash/isEqual';

import { makeExchange } from "../actions/currencyAction";
import {formatWithFloating,  getNewCurrencyIndex, validateCurrencyNumber} from "../utils/currency";

import ExchangeInput from "../components/ExchangeInput";
import ExchangeSlide from '../components/ExchangeSlide';
import ExchangeRotate from "../components/ExchangeRotate";
import ExchangeButton from "../components/ExchangeButton";
import RatioDisplay from "../components/RatioDisplay";

const DIR = {
    TARGET : 'TARGET',
    SOURCE: 'SOURCE'
};

const CLEARED_STATE = {
    targetValue: '0.00',
    sourceValue: '0.00',
};

export class ExchangeContainer extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            exchangeWarning: '',
            exchangeDisabled: true,
            targetIndexCurrency: 0,
            sourceIndexCurrency: 1,
            
            targetValue: '0.00',
            sourceValue: '0.00'
        };

        this.exchange = this.exchange.bind(this);
        this.checkValueSource = this.checkValueOnBlur.bind(this, 'sourceValue');
        this.checkValueTarget = this.checkValueOnBlur.bind(this, 'targetValue');
    }

    componentDidUpdate(prevProps){
        if(!isEqual(prevProps.currencies,this.props.currencies)) {
            this.calculateCurrencyValues(DIR.SOURCE, this.state.sourceValue);
        }
    }
    
    getRate(source : number, target : number) {
        const { currencies } = this.props;
        return currencies[target].rates[currencies[source].code];
    }

    exchange() {
        const { currencies, makeExchange } = this.props;
        const { targetIndexCurrency, sourceIndexCurrency, targetValue, sourceValue} = this.state;
        this.setState({
            ...CLEARED_STATE,
            exchangeDisabled: true
        });
        return makeExchange(currencies[sourceIndexCurrency], currencies[targetIndexCurrency], sourceValue, targetValue);
    }
    
    
    calculateCurrencyValues(direction, value) {
        if (value === '') {
            this.setState(CLEARED_STATE);
        } else {
            const { targetIndexCurrency, sourceIndexCurrency } = this.state;

            if (direction === DIR.SOURCE) {
                const recalculatedValue = new Big(value)
                    .times(this.getRate(targetIndexCurrency, sourceIndexCurrency))
                    .toFixed(2);

                this.setState({
                    targetValue: formatWithFloating(recalculatedValue)
                });
                this.checkIfCanExchange(value);
            } else {
                const recalculatedValue = new Big(value)
                    .times(this.getRate(sourceIndexCurrency, targetIndexCurrency))
                    .toFixed(2);
                this.setState({
                    sourceValue: formatWithFloating(recalculatedValue)
                });
                this.checkIfCanExchange(recalculatedValue);

            }
        }
    };

    checkValueOnBlur(field,value) {
        this.checkIfCanExchange(this.state.sourceValue);
        if(parseFloat(value) === 0) {
            this.setState();
        } else {
            this.setState({
                [field]: formatWithFloating(value)
            });
        }
    }

    checkIfCanExchange(value : string) {
        const { sourceIndexCurrency } = this.state;
        const { accounts, currencies } = this.props;
        const balance = accounts[currencies[sourceIndexCurrency].code].balance;
        this.setState({
            exchangeWarning: parseFloat(balance) < parseFloat(value) ? 'There are insufficient funds to make exchange' : '',
            exchangeDisabled: parseFloat(value) === 0 || parseFloat(balance) < parseFloat(value),
        });
    }

    onValueChange(direction : string, value : string) {
        if (validateCurrencyNumber(value)) {
            const formattedValue = value !== '' ? value : '';
            if(direction === DIR.SOURCE) {
                this.setState({
                    sourceValue: formattedValue
                });
            } else {
                this.setState({
                    targetValue: formattedValue
                });
            }
            this.calculateCurrencyValues(direction, value);
        }
    };

    onSlideTarget(value : string) {
        const { targetIndexCurrency, sourceIndexCurrency } = this.state;
        const { currencies } = this.props;

        let newValueTmp = getNewCurrencyIndex(targetIndexCurrency, value, currencies, sourceIndexCurrency); 
        this.setState({
            targetIndexCurrency: newValueTmp 
        }, () => {
            this.calculateCurrencyValues(DIR.TARGET, this.state.targetValue);
        });
    }

    onSlideSource(value : string) {
        const { targetIndexCurrency, sourceIndexCurrency } = this.state;
        const { currencies } = this.props;

        let newValueTmp = getNewCurrencyIndex(sourceIndexCurrency, value, currencies, targetIndexCurrency); 

        this.setState({
            sourceIndexCurrency: newValueTmp 
        }, () => {
            this.calculateCurrencyValues(DIR.SOURCE, this.state.sourceValue);
        });
    }

    onRevertCurrencies() {
        this.setState({
            targetIndexCurrency: this.state.sourceIndexCurrency,
            sourceIndexCurrency: this.state.targetIndexCurrency,
            sourceValue: this.state.targetValue
        }, () => {
            this.calculateCurrencyValues(DIR.SOURCE, this.state.sourceValue);
        });
        
    }

    render() {
        const { currencies, accounts } = this.props;
        const { targetIndexCurrency, sourceIndexCurrency, exchangeDisabled, exchangeWarning } = this.state;

        if (currencies.length > 2) {
            const ratio = this.getRate(targetIndexCurrency, sourceIndexCurrency);
            return (
                <div className="exchange-container container">
                        <ExchangeSlide
                            onLeftClick={this.onSlideSource.bind(this, -1)}
                            onRightClick={this.onSlideSource.bind(this, 1)}
                        >
                            <ExchangeInput
                                currency={currencies[sourceIndexCurrency]}
                                balance={accounts[currencies[sourceIndexCurrency].code].balance}
                                onValueChange={this.onValueChange.bind(this, DIR.SOURCE)}
                                onBlur={this.checkValueSource}
                                value={this.state.sourceValue}
                            />
                        </ExchangeSlide>
                        <RatioDisplay ratio={ratio} ratioCharFrom={currencies[sourceIndexCurrency].char}
                                      ratioCharTo={currencies[targetIndexCurrency].char}/>
                        
                        <ExchangeRotate onClick={this.onRevertCurrencies.bind(this)}/>
                        <ExchangeSlide
                            onLeftClick={this.onSlideTarget.bind(this, -1)}
                            onRightClick={this.onSlideTarget.bind(this, 1)}
                        >
                            <ExchangeInput
                                currency={currencies[targetIndexCurrency]}
                                balance={accounts[currencies[targetIndexCurrency].code].balance}
                                onValueChange={this.onValueChange.bind(this, DIR.TARGET)}
                                onBlur={this.checkValueTarget}
                                value={this.state.targetValue}
                            />
                        </ExchangeSlide>
                        <ExchangeButton 
                            label={'Exchange!'}
                            onClick={this.exchange}
                            isDisabled={exchangeDisabled}
                            disabledMsg={exchangeDisabled}
                        />
                        <div className="text-center">{exchangeWarning}</div>
                </div>
            )
        }
        return null;
    }
}

const mapStateToProps = ({wallet, currency}) => {
    return {
        accounts: wallet.get('accounts').toJS(),
        currencies: currency.get('entries').toJS()
    }
};

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        makeExchange
    }, dispatch);
}


export default connect(mapStateToProps, mapDispatchToProps)(ExchangeContainer);

