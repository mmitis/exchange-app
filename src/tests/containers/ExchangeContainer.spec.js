import { ExchangeContainer } from './../../containers/ExchangeContainer';
import { mount } from 'enzyme';
import React from "react";

describe('Component: ExchangeContainer', () => {
    const props = {
        accounts: {
            PLN : { balance : 100 },
            USD : { balance : 100 }
        },
        currencies: [
            { code: 'PLN', char: 'zł', rates: { USD: 2 }},
            { code: 'USD', char: '$', rates: { PLN: 0.5}}
        ],
        makeExchange: jest.fn()
    }

    it('test simple render', async () => {
        const container = mount(<ExchangeContainer
            {...props}
        />);
        
        expect(container).toMatchSnapshot();
    });

    it('should simple calculate ratio', async () => {
        const container = mount(<ExchangeContainer
            {...props}
        />);
        const instance = container.instance();
        instance.calculateCurrencyValues('SOURCE', 10);
        container.update();
        expect(instance.state.targetValue).toEqual('5.00');
        instance.calculateCurrencyValues('TARGET', 10);
        container.update();
        expect(instance.state.sourceValue).toEqual('20.00');

    });

    it('should call properly exchange', async () => {
        const container = mount(<ExchangeContainer
            {...props}
        />);
        const instance = container.instance();
        instance.exchange();
        expect(props.makeExchange).toHaveBeenCalledWith({"char": "$", "code": "USD", "rates": {"PLN": 0.5}}, 
        {"char": "zł", "code": "PLN", "rates": {"USD": 2}}, "0.00", "0.00");
    });

    it('should get rate', async () => {
        const container = mount(<ExchangeContainer
            {...props}
        />);
        const instance = container.instance();
        expect(instance.getRate(0,1)).toEqual(0.5);
        expect(instance.getRate(1,0)).toEqual(2);
    });

    it('should check if can get exchange', async () => {
        const container = mount(<ExchangeContainer
            {...props}
        />);
        const instance = container.instance();
        instance.checkIfCanExchange('120.00');
        container.update();
        expect(instance.state.exchangeDisabled).toEqual(true);
        instance.checkIfCanExchange('90.00');
        container.update();
        expect(instance.state.exchangeDisabled).toEqual(false);
    });

    it('should revert currencies', async () => {
        const container = mount(<ExchangeContainer
            {...props}
        />);
        const instance = container.instance();
        instance.setState({
            targetValue: 5,
            sourceValue: 3
        });
        container.update();
        instance.onRevertCurrencies();
        container.update();
        expect(instance.state).toEqual({
            "exchangeDisabled": false,
            "exchangeWarning": "",
            "sourceIndexCurrency": 0,
            "sourceValue": 5,
            "targetIndexCurrency": 1,
            "targetValue": "10.00",
        })
    });

});
