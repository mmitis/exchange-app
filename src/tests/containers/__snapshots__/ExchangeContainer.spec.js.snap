// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ExchangeContainer test simple render 1`] = `ReactWrapper {}`;
